include/deal.II/base/mpi.h

```diff
@@ -708,6 +709,416 @@ namespace Utilities
+      std::vector<std::vector<T2>> recv_buffers;
+      std::vector<MPI_Request>     send_and_recv_buffers;
+
+      // data structures to answer requests
+      std::vector<std::vector<T2>> requests_buffers;
+      std::vector<MPI_Request>     requests_answers;
+#endif
+
+      /**
+       * The ith request message from another rank has been received: process
+       * the request and send an answer.
+       */
+      void
+      process_requests(int index);
+
+      /**
+       * Start to send all requests via ISend and post IRecvs for the incoming
+       * answer messages.
+       */
+      unsigned int
+      start_communication();
+
+      /**
+       * After all answers have been exchanged, the MPI data structures can be
+       * freed and the received answers can be processed.
+       */
+      void
+      clean_up_and_end_communication();
+    };
+
+    /**
+     * A class which delegates its task to other ConsensusAlgorithm
+     * implementations depending on the number of processes in the
+     * MPI communicator. It uses for small number of processes
+     * ConsensusAlgorithm_PEX and for large number of processes
+     * ConsensusAlgorithm_NBX. The threshold depends if the program is
+     * compiled in debug or release mode.
+     *
+     * @author Peter Munch, 2019
+     */
+    template <typename T1, typename T2>
+    class ConsensusAlgorithmSelector : public ConsensusAlgorithm<T1, T2>
+    {
+    public:
+      /**
+       * Constructor.
+       *
+       * @param process Process to be run during consensus algorithm.
+       * @param comm MPI Communicator.
+       */
+      ConsensusAlgorithmSelector(ConsensusAlgorithmProcess<T1, T2> &process,
+                                 const MPI_Comm &                   comm);
+
+      /**
+       * Destructor.
+       */
+      virtual ~ConsensusAlgorithmSelector();
+
+      /**
+       * Run consensus algorithm. The function call is delegated to another
+       * ConsensusAlgorithm implementation.
+       */
+      void
+      run() override;
+
+    private:
+      // Pointer to the actual ConsensusAlgorithm implementation.
+      std::shared_ptr<ConsensusAlgorithm<T1, T2>> consensus_algo;
+    };
+
+    /**
+     * Given a partitioned index set space, compute the owning MPI process rank
+     * of each element of a second index set according to the partitioned index
+     * set. A natural usage of this function is to compute for each ghosted
+     * degree of freedom the MPI rank of the process owning that index.
+     *
+     * One might think: "But we know which rank a ghost DoF belongs to based on
+     * the subdomain id of the cell it is on". But this heuristic fails for DoFs
+     * on interfaces between ghost cells with different subdomain_ids, or
+     * between a ghost cell and an artificial cell."
+     *
+     * The first argument passed to this function, @p owned_indices, must
+     * uniquely partition an index space between all processes.
```

Comment:
Side question: Must it be a partition that is 1-to-1 **and onto** or can the index sets exclude some indices? For example, think of the case when you have a distributed set of indices and you additional skip the pressure dofs from a velocity/pressure dof handler pair. I think the algorithm should be able to handle this, but I read "uniquely partition" as being bijective. I think you only want the owned index sets to be injective (1-to-1 without being surjective) where no index is contained in more than one set (but it could be in none).

Reply:
At the moment, the relation indeed has to be bijective. The reason is this line in the dictionary:
```cpp
while (this->local_size != dic_local_rececived)
```
It will only terminate iff it index set if uniquely partitioned. If this is, too, restrictive we could pass some kind of additional data.

Reply:
@kronbichler: If the bijective assumption is too restrictive, I will extend the functionality of this class in a next PR.